src/tests/conformance/query/peer_shard_info.rs

```diff
@@ -24,3 +35,86 @@ async fn c006_TM_GET_PEER_SHARD_INFO_V2_node_should_not_query_for_shard_info_if_
         |m: &BinaryMessage| matches!(&m.payload, Payload::TmGetPeerShardInfoV2(..));
     perform_response_test(TestConfig::default().with_handshake(false), &response_check).await;
 }
+
+#[tokio::test]
+#[allow(non_snake_case)]
+async fn c011_TM_GET_PEER_SHARD_INFO_V2_node_should_relay_shard_info() {
+    // ZG-CONFORMANCE-011
+    for key_type in KEY_TYPES {
+        check_relay_for_key_type(*key_type, 2).await;
+    }
+}
+
+#[tokio::test]
+#[should_panic]
+#[allow(non_snake_case)]
+async fn c012_TM_GET_PEER_SHARD_INFO_V2_node_should_not_relay_shard_info_with_invalid_key_type() {
+    // ZG-CONFORMANCE-012
+    check_relay_for_key_type(0x42, 2).await;
+}
+
+#[tokio::test]
+#[should_panic]
+#[allow(non_snake_case)]
+async fn c013_TM_GET_PEER_SHARD_INFO_V2_node_should_not_relay_shard_info_when_relays_equals_zero() {
+    // ZG-CONFORMANCE-013
+    check_relay_for_key_type(KEY_TYPES[0], 0).await;
+}
+
+#[tokio::test]
+#[should_panic]
+#[allow(non_snake_case)]
+async fn c014_TM_GET_PEER_SHARD_INFO_V2_node_should_not_relay_shard_info_when_relays_above_limit() {
+    // ZG-CONFORMANCE-014
+    check_relay_for_key_type(KEY_TYPES[0], 4).await;
+}
+
+async fn check_relay_for_key_type(key_type: u8, relays: u32) {
+    // Create node.
+    let target = TempDir::new().expect("unable to create TempDir");
+    let mut node = Node::builder()
+        .log_to_stdout(false)
+        .start(target.path(), NodeType::Stateless)
+        .await
+        .expect("unable to start rippled node");
+
+    // Create two synthetic nodes and connect them to rippled.
+    let synth_node1 = SyntheticNode::new(&Default::default()).await;
+    synth_node1
+        .connect(node.addr())
+        .await
+        .expect("unable to connect");
+    let mut synth_node2 = SyntheticNode::new(&Default::default()).await;
+    synth_node2
+        .connect(node.addr())
+        .await
+        .expect("unable to connect");
+
+    // Create a dummy key with the specified key type.
+    let mut key = vec![key_type];
+    key.extend(vec![0x1; 32]);
+    let public_key = TmPublicKey { public_key: key };
+
+    // Create payload with given key and relays.
+    let payload = Payload::TmGetPeerShardInfoV2(TmGetPeerShardInfoV2 {
+        peer_chain: vec![public_key.clone()],
+        relays,
+    });
+    // Send a message from the first synthetic node.
+    synth_node1
+        .unicast(node.addr(), payload)
+        .expect("unable to send message");
+
+    // Ensure that the second synthetic node receives relayed message.
+    // Check for public key and reduced `relays` number.
+    let check = |m: &BinaryMessage| {
+        matches!(&m.payload, Payload::TmGetPeerShardInfoV2(TmGetPeerShardInfoV2{peer_chain, relays: received_relays})
+          if !peer_chain.is_empty() && peer_chain[0] == public_key && received_relays == &(relays - 1))
```

Comment:
```suggestion
          if !peer_chain.is_empty() && peer_chain.get(0) == Some(public_key) && received_relays == &(relays.saturating_sub(1)))
```
panic safety

Reply:
also, shouldn't `received_relays == &(relays - 1)` be `received_relays == (relays - 1)` instead (without the `&`)?

Reply:
Then it should instead be `*received_relays == relays.saturating_sub(1)`

Reply:
Done